refactor(cast-details): improve UI rendering and scrolling behavior

skaik-mo opened the pull request:
Key Changes:

- Conditional UI Rendering:

  - Hide the gallery section when there are no images.
  - Hide the actor's birthday and place of birth if data is missing.
  - Hide the info section when the actor’s description is blank.

- Layout & UX Improvements:

  - Replaced LazyColumn with Column and verticalScroll since the content is relatively short and doesn't require lazy rendering.
  - Improved spacing and padding for better visual consistency.
  - Applied dynamic bottom padding to ensure content isn't overlapped by the system navigation bar, adapting based on content height.

- Code Refactoring:
  - Cleaned up and simplified mapping logic for UI components.


- Refactor: Person Image Mapping & De-duplication

  - Updated mapToPerson() to accept a List<ProfileDto> instead of the full PersonProfileImageDto object.
  - Modified PersonRepositoryImpl to drop the first image in the list, since it often duplicates the main profile image.
  - Added a toImageUrl() extension function to ProfileDto for simplified and consistent image URL construction.
  - Removed the obsolete PersonProfileImageDto.toImageList() extension.

skaik-mo committed:
refactor: replace LazyColumn with Column and enhance scrolling behavior in CastDetailsScreen

skaik-mo committed:
refactor:
- implement animated visibility for loading state in CastDetailsScreen
- replace LazyColumn with Column and enhance scrolling behavior in CastDetailsScreen

skaik-mo committed:
Merge branch 'develop' into fix/cast-details-screen

* develop: (78 commits)
  Refactor: Update GetMovieReviewsUseCaseTest to use common Review entity
  add test cases to search by name
  change search series by name to return sorted series
  add count feature to series also
  update mappers
  delete movie and series genre
  delete migration
  add one genre class to movie and series
  Migrate SeriesReview to Review in fakeSeries
  feat: display movie reviews in movie details screen
  Refactor ReviewsScreen to use ReviewUI model
  Refactor: Simplify getMovieDetails method in MoviesRepositoryImpl
  Refactor MovieReviewMapper to use a generic Review entity
  Refactor: Use UI models in MovieDetails ViewModel
  Refactor: Update Review entity path in use cases
  Refactor series reviews mapper The `parseData` function has been updated to return `LocalDateTime` instead of `LocalDate`.
  Refactor: Adjust padding in AddToCollectionContent and update StarCastSection
  Refactor: Use generic Review entity
  Refactor: Convert MovieUi to/from Entity mapping to extension functions
  feat: Add review and crew UI ,cast models
  ...

skaik-mo committed:
Merge branch 'develop' into fix/cast-details-screen

* develop: (186 commits)
  Remove Firebase debug deployment workflow
  resolve conflicts in viewModel
  replace String to @Serializable in recommended series route
  refactor call function navigate to series details
  Fix season screen
  Remove validation files and logic
  Refactor: Update error message and use method references
  refactor primary button composable
  Add unknown error string resource
  Refactor: Remove Ktor and replace with Retrofit for network requests
  Refactor: Migrate User Remote Data Source to Retrofit/OkHttp
  Add ExploreApiServiceRetrofit and its implementation
  feat: Implement Retrofit service and data source for person API
  feat: Implement Retrofit-based remote data source for series
  Refactor: Implement Retrofit for network calls and convert movie
  fix get genre
  fix navigation between recommended series and series details
  fix navigation to recommended series screen
  add transition animation
  fix: remove redundant size assignment in CastItem and StarCastSection
  ...

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsScreen.kt

skaik-mo committed:
Refactor: Replace LazyColumn with Column in CastDetailsScreen

The CastDetailsScreen has been refactored to use a `Column` with a `verticalScroll` modifier instead of a `LazyColumn`.

This change was made because the screen content is not excessively long, and a simple `Column` is more appropriate and potentially more performant for this use case.

Additionally, dynamic bottom padding has been implemented to ensure content is not obscured by the navigation bar, adjusting based on whether the content fits within the screen height.

skaik-mo committed:
fix cast details ui

- Hide gallery section if there are no images.
- Hide actor's birthday and place of birth if they are not available.
- Hide info section if description is blank.
- Refactor mapping logic.
- Adjust spacing in cast details screen.

skaik-mo committed:
Refactor person image mapping and prevent duplication

- Updated the `mapToPerson` function to accept a `List<ProfileDto>` instead of `PersonProfileImageDto` for images.
- Modified the `PersonRepositoryImpl` to drop the first image from the fetched profile images, as it is typically redundant with the main profile image.
- Simplified image URL generation by creating a `toImageUrl` extension function for `ProfileDto`.
- Removed the `PersonProfileImageDto.toImageList()` extension function.

mohannadahmed00 commented on repository/media/src/main/java/com/giraffe/media/person/PersonRepositoryImpl.kt:
```diff
@@ -88,17 +87,14 @@ class PersonRepositoryImpl(
+            mapToPerson(
+                personId = personId,
+                details = details.await(),
+                // The first image is typically the profile image already included in person details.
```
remove these comments

poula22 replied:
@skaik-mo

skaik-mo replied:
This comment is added to clarify for other developers why the first item is being dropped.

poula22 reviewed (dismissed):
Great Job👏👏

skaik-mo committed:
Merge branch 'develop' into fix/cast-details-screen

* develop: (54 commits)
  Refactor: Remove `UnauthorizedException` and its handling
  changed the file name to camel case
  Refactor: Consolidate network-related exceptions into `NoInternetException`. - Replacing `NetworkException`, `TimeoutException`, and `ServerErrorException` with a single `NoInternetException`. - Removing `InvalidRequestMethodException` and `InfrastructureException` as they are now covered by `NoInternetException` or other existing exceptions. - Updating `SafeCall` to map various data layer exceptions (like `ServerDataException`, `RequestTimeoutDataException`) to the new `NoInternetException`.
  Refactor: Simplify UserApiServiceRetrofit instantiation
  Refactor: Update person search API to support pagination
  Refactor: Rename Retrofit builder instances for consistency
  Refactor: Define constants for Retrofit API endpoints and parameters
  Refactor: Network module configuration
  fix: update searchByName method to include pagination and change response type
  it will be implemented.
  add get top rated series use case test.
  add get recently released series use case test.
  add get popularity series use case test.
  add get top rated series use case.
  add get recently released series use case.
  add get popularity series use case.
  add the functions get popularity series, get recently released series, and get top rated series to the series repository.
  move series use cases to series package.
  remove home repository.
  Refactor: Correct typo in GalleryRoute data class and update navigation functions
  ...

# Conflicts:
#	repository/media/src/main/java/com/giraffe/media/person/mapper/Mapper.kt

poula22 commented on repository/media/src/main/java/com/giraffe/media/person/mapper/Mapper.kt:
```diff
@@ -90,8 +112,7 @@ fun PersonSocialMediaDto.toEntity(): PersonSocialMediaLinks = PersonSocialMediaL
-    profiles.map { BASE_IMAGE_URL + it.filePath }
+fun ProfileDto.toImageUrl(): String = BASE_IMAGE_URL + filePath
 
 fun String?.prependIfNotBlank(prefix: String): String? =
     this?.takeIf { it.isNotBlank() }?.let { prefix + it }
```
This function is applied to all nullable strings as extension function, which doesn't make sense here. It should add prefix only to String which is not null

skaik-mo replied:
You are right, I will fix this in a separate PR.

poula22 reviewed (dismissed):
Just one comment. Great job👏👏

K-M-9 commented on designsystem/src/main/java/com/giraffe/designsystem/composable/InfoSection.kt:
```diff
@@ -19,6 +19,7 @@ fun InfoSection(
     descriptionLimits: Int = 100,
     spaceBetween: Int = 8
 ) {
+    if (description.isBlank()) return
```
It is `better` to be
`if (description.isNotBlank()){
Column{}
}`

skaik-mo committed:
Merge branch 'develop' into fix/cast-details-screen

* develop: (21 commits)
  add navigate to series details
  refactor some component
  resolve conflict in nav graph
  refactor star cast section
  refactor animation in series details
  merge develop and resolve conflict
  refactor viewmodel to make functions more clean
  refactor animation in cast credits
  fix(tests): remove unnecessary single() call from getSearchKeywords verification
  fix(api): update search API response type to SearchPersonResponse
  fix: implement pagination for searchByName and update cache management for search keywords
  merge develop
  refactor series details ui
  refactor viewmodel in series details
  fix navigation and route in series details
  formating release year
  create ui screen for cast credit
  create transition animation between column and vertical grid
  add navigate from cast details to cast credit add cast credit route in details nav graph
  create cast credit route
  ...

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsScreen.kt
#	repository/media/src/main/java/com/giraffe/media/person/mapper/Mapper.kt

skaik-mo committed:
add animated visibility to sections

This commit introduces `AnimatedVisibility` to the `InfoSection`, `PosterListSection`, and `GallerySection` composables. This enhancement ensures that these sections only appear if they have content, providing a smoother user experience.

Additionally, the `prependIfNotBlank` function in `repository/media` has been updated to use `isNullOrBlank` for improved null safety and clarity.

skaik-mo committed:
add navigation from cast details to movie and series details

- Refactored `PosterListSection` to take a list of `Poster` objects instead of a list of `Movie` objects.
- Added navigation from the cast details screen to the gallery screen.
- Added navigation from the cast details screen to the movie details screen.
- Added navigation from the cast details screen to the series details screen.
- Updated `PosterItemVertically` to take a `Poster` object instead of a `Movie` object.
- Updated `CastDetailsEffect` to include `NavigateToMovieDetails` and `NavigateToSeriesDetails` effects.
- Updated `CastDetailsViewModel` to send `NavigateToMovieDetails` and `NavigateToSeriesDetails` effects when a poster is clicked.
- Updated `CastDetailsScreen` to handle `NavigateToMovieDetails` and `NavigateToSeriesDetails` effects.

skaik-mo committed:
Refactor CastDetailsViewModel to use SavedStateHandle

The CastDetailsViewModel now retrieves the `personId` from the `SavedStateHandle` instead of receiving it as a constructor parameter. This simplifies dependency injection and aligns with Android Jetpack best practices for handling screen state.

The following changes were made:
- Updated `viewModelModule` to provide `SavedStateHandle` to `CastDetailsViewModel`.
- Modified `CastDetailsViewModel` to accept `SavedStateHandle` and extract `personId` using `toRoute<CastDetailsRoute>()`.
- Removed the `personId` parameter from `CastDetailsScreen` and `castDetailsRoute`.

skaik-mo committed:
Update error handling in CastDetails

- Added `errorMessage` to `CastDetailsUiState` to display errors in the UI.
- Removed `CastDetailsEffect.Error` as errors are now handled via `CastDetailsUiState`.
- Modified `CastDetailsViewModel` to update `errorMessage` in `CastDetailsUiState` upon error.
- Updated `CastDetailsScreen` to display an error message at the bottom when `errorMessage` is present in the state.

K-M-9 commented on presentation/details/src/main/java/com/giraffe/details/screens/castDetails/CastDetailsRoute.kt:
```diff
@@ -14,18 +13,19 @@ internal fun NavController.navigateToCastDetails(id: Int) {
 
 internal fun NavGraphBuilder.castDetailsRoute(
-    navController: NavController,
+    navigateToGallery: (String, List<String?>) -> Unit,
```
Avoid passing `lists` through `navigation` arguments due to `size` limitations.

skaik-mo committed:
Add Gallery screen for person details

This commit introduces the following changes:

- Creates a new Gallery screen to display images of a person.
- Adds a `GalleryViewModel` to manage the UI state of the Gallery screen.
- Implements the `GetPersonImagesUseCase` to fetch person images from the repository.
- Updates the navigation to include the Gallery screen.
- Adds a `LoadingView` component to display a loading indicator.
- Refactors the `CastDetailsEffect` to pass the `personId` instead of a list of image URLs to the Gallery screen.

skaik-mo committed:
Merge branch 'develop' into fix/cast-details-screen
